feat: add serviceName option to ThanosRuler and AlertManager CRD

trouaux opened the pull request:
## Description

Follow up of https://github.com/prometheus-operator/prometheus-operator/pull/6687

Thanos CRD
  - new ServiceName field
  - updated statefulset.go and operator.go
  - updated e2e

K8sUtils
  - moved EnsureCustomGoverningService 

Prometheus Agent + Server 
- updated statefulset.go
- operator.go using k8sutils


AM 
  - new ServiceName field
  - updated statefulset.go and operator.go
  - updated e2e

Closes 
[#7317](https://github.com/prometheus-operator/prometheus-operator/issues/7317)
https://github.com/prometheus-operator/prometheus-operator/pull/7329


## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
add serviceName option to ThanosRuler and Alertmanager CRDs
```

trouaux commented:
@simonpasquier first draft of the PR

does it makes sense to have `EnsureCustomGoverningService` in k8sutils ?

heliapb commented on pkg/apis/monitoring/v1/thanos_types.go:
```diff
@@ -140,6 +140,14 @@ type ThanosRulerSpec struct {
+	// If empty, the operator will create and manage a headless service named `thanos-ruler-operated` for ThanosRuler resources.
+	// When deploying multiple ThanosRuler resources in the same namespace, it is recommended to specify a different value for each.
+	// See https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id for more details.
+	// +optional
```
Hi @trouaux probably better to add `// +kubebuilder:validation:MinLength=1 `to service name, similar to https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/apis/monitoring/v1/prometheus_types.go#L881

simonpasquier commented on pkg/prometheus/server/statefulset.go:
```diff
@@ -332,8 +332,13 @@ func makeStatefulSetSpec(
+
 	spec := appsv1.StatefulSetSpec{
-		ServiceName: governingServiceName,
+		ServiceName: serviceName,
```
(nit)
```suggestion
		ServiceName: ptr.Deref(cpf.ServiceName, governingServiceName),
```

simonpasquier commented on pkg/prometheus/server/statefulset.go:
```diff
@@ -332,8 +332,13 @@ func makeStatefulSetSpec(
 
+	serviceName := governingServiceName
+	if cpf.ServiceName != nil {
+		serviceName = *cpf.ServiceName
```
Good catch! We missed in #6687 that the service name needs to be propagated to the statefulset spec :fearful: 

cc @mviswanathsai

mviswanathsai replied:
Whew! good catch indeed.

simonpasquier commented on pkg/thanos/operator.go:
```diff
@@ -794,3 +801,9 @@ func newTLSAssetSecret(tr *monitoringv1.ThanosRuler, config Config) *v1.Secret {
 	return s
 }
+
+func makeSelectorLabels(name string) map[string]string {
```
I would include the labels from here and modify makeStatefulSetSpec() to call makeSelectorLabels() so we have one source of truth:

https://github.com/prometheus-operator/prometheus-operator/blob/838b6a701c0cbd5ff843f7cbcfef7546b4c50eb1/pkg/thanos/statefulset.go#L381-L384

simonpasquier commented on pkg/k8sutil/k8sutil.go:
```diff
@@ -566,3 +567,32 @@ func UpdateDNSPolicy(podSpec *v1.PodSpec, dnsPolicy *monitoringv1.DNSPolicy) {
+//
+// If the ServiceName is specified and a service with the same name exists in the same namespace as the
+// resource, ensure that the custom governing service's selector matches the
+// statefulsets.
```
(nit) not your responsibility but the function only gets a label map.
```suggestion
// labels.
```

simonpasquier commented on pkg/k8sutil/k8sutil.go:
```diff
@@ -566,3 +567,32 @@ func UpdateDNSPolicy(podSpec *v1.PodSpec, dnsPolicy *monitoringv1.DNSPolicy) {
+// statefulsets.
+// If it is not selected, fail the reconciliation
+// Warning: the function will panic if the resource's ServiceName is nil..
+func EnsureCustomGoverningService(ctx context.Context, namespace string, serviceName string, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
```
(nit)
```suggestion
func EnsureCustomGoverningService(ctx context.Context, namespace string, serviceName string, svcClient clientv1.ServiceInterface, labels map[string]string) error {
```

trouaux replied:
this change would conflict with `"k8s.io/apimachinery/pkg/labels"`

trouaux commented:
@simonpasquier I addressed your comments, not sure how e2e for alertmanager broke
`/metrics: the server is currently unable to handle the request (get pods https:alertmanager-reload-config-0:)`

here the last log from alertmanager-reload-config-0
`level=info ts=2025-02-18T10:58:10.996019267Z caller=reloader.go:548 msg="Reload triggered" cfg_in=/etc/alertmanager/config/alertmanager.yaml.gz cfg_out=/etc/alertmanager/config_out/alertmanager.env.yaml cfg_dirs= watched_dirs="/etc/alertmanager/config, /etc/alertmanager/secrets/alertmanager-templates-reload-config, /etc/alertmanager/configmaps/alertmanager-templates-reload-config"
`


should we rebase https://github.com/prometheus-operator/prometheus-operator/pull/7329 into this PR ?

trouaux commented:
@jkroepke I updated my PR with changes necessary for AM. Will be easier to have both in 1 PR
https://github.com/prometheus-operator/prometheus-operator/pull/7329

simonpasquier reviewed (approved):
thanks!
